### PR TITLE
Add validation

### PR DESCRIPTION
**Pull request by mikefaut**
We want to have validation on the proxy so that clients can be dumb about publishing.

Currently messages will dual publish if they are invalid. So they will still publish the invalid message but will also publish an invalid message event to a separate channel.

By default validation is disabled.

### CODE REVIEW

File: src/pubsub.js [hunk 1]
```diff
@@ -3,9 +3,59 @@
...
+let publish = async (channel, messages) => {
+  let errors = [];
+  let validMessages = [];
+  let invalidMessages = [];
+  // Map all messages to validate and publish each individually.
+  await Promise.all(messages.map((message)=>{
+    return validator.validate(message).then(()=>{
+      validMessages.push(message);
+      // If a message is valid then publish it.
+      prom.messageCount.inc({state: 'valid'})
```
Comment (geyer-za) [inline src/pubsub.js:25]: what happens here is say prom is down? would this queue locally and try later? or fail hard here?

Comment (geyer-za) [reply src/pubsub.js:25]: maybe wrap this in a try catch like you did below?

Comment (mikefaut) [reply src/pubsub.js:25]: prom doesn't implement queues or pushing to prom. Its a variable/state that it reads. So its just increasing a variable by 1.
Then the scraper comes along and reads that state. 
So if prom is down this application has no idea that prom is down. (only that the metrics endpoint isn't getting hit)

File: src/app.js [hunk 2]
```diff
@@ -14,12 +14,10 @@ let prom = require('./prometheus');
 
 let logger = require('./logger');
 let pubsub = require('./pubsub');
-let connection = pubsub.connection();
 let queue = require('./queue');
 
 // this flag indicates whether or not we'll accept messages on the POST end-point
 // when a SIGTERM is received, we set this flag to false and throw back a 503
```
Comment (nicja) [inline src/app.js]: Should probably remove this comment too

**Review by nicja — COMMENTED** [review]
Not too much context, but LGTM

**Review by nicja — APPROVED** [review]

File: src/app.js (hunk 2 above)
Comment (mikefaut) [reply src/app.js]: Good spot. Will get rid of it

File: src/app.js [hunk 3]
```diff
@@ -90,20 +92,28 @@ app.use((err, req, res, next) => {
...
 
-  logger.info('Processing remaining jobs on queue');
-  queue.start(() => {
-    logger.info('Closing express server socket');
-    app.server.close(() => {
-      process.exit(0);
-    });
-  });
+  logger.info('Closing express server socket');
+  // When the HTTP server closes we want to empty the job queue.
```
Comment (quintonparker) [inline src/app.js:100]: 👍

File: src/pubsub.js (hunk 1 above)
Comment (quintonparker) [reply src/pubsub.js:25]: 👍 +1

File: src/pubsub.js [hunk 4]
```diff
@@ -3,9 +3,65 @@
...
+      return connection.publish(channel, message);
+    }).catch((error) => {
+      if(error.name == 'ValidationError') {
+        logger.warn('ValidationError: '+channel+' - '+error.event.errors);
```
Comment (quintonparker) [inline src/pubsub.js]: es6 template literals #justsayin

File: src/validator.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
+const config = require('./config');
+const logger = require('./logger');
+//  By default validation is always happy
+let validate = () => {
+  return Promise.resolve();
```
Comment (quintonparker) [inline src/validator.js:5]: clever

**Review by quintonparker — APPROVED** [review]
looks good to me 🤷‍♂️

**Pull request merged**